# _Tasca 5, Sprint 3:_ *Dataframe Structures* 

## Exercise 1

Download the "Airlines Delay: Airline on-time statistics and delay causes" data set and upload it to a Dataframe pandas. Explore the data it contains, and keep only the columns you consider relevant.

In [1]:
# importing libraries
import pandas as pd
import numpy as np
pd.set_option ('display.max_columns', None)
#pd.set_option ('display.max_rows', None)

In [2]:
pd.options.display.float_format = '{:.0f}'.format
filename = '../Data/DelayedFlights.csv'

data = pd.read_csv(filename, delimiter=',')

print('Shape of the data is', data.shape)
data.describe()

Shape of the data is (1936758, 30)


Unnamed: 0    Year   Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
count     1936758 1936758 1936758     1936758    1936758  1936758     1936758   
mean      3341651    2008       6          16          4     1519        1467   
std       2066065       0       3           9          2      450         425   
min             0    2008       1           1          1        1           0   
25%       1517452    2008       3           8          2     1203        1135   
50%       3242558    2008       6          16          4     1545        1510   
75%       4972467    2008       9          23          6     1900        1815   
max       7009727    2008      12          31          7     2400        2359   

       ArrTime  CRSArrTime  FlightNum  ActualElapsedTime  CRSElapsedTime  \
count  1929648     1936758    1936758            1928371         1936560   
mean      1610        1634       2184                133             134   
std        548         465       1945                 72              71   
min          1           0          1                 14             -25   
25%       1316        1325        610                 80              82   
50%       1715        1705       1543                116             116   
75%       2030        2014       3422                165             165   
max       2400        2400       9742               1114             660   

       AirTime  ArrDelay  DepDelay  Distance  TaxiIn  TaxiOut  Cancelled  \
count  1928371   1928371   1936758   1936758 1929648  1936303    1936758   
mean       108        42        43       766       7       18          0   
std         69        57        53       574       5       14          0   
min          0      -109         6        11       0        0          0   
25%         58         9        12       338       4       10          0   
50%         90        24        24       606       6       14          0   
75%        137        56        53       998       8       21          0   
max       1091      2461      2467      4962     240      422          1   

       Diverted  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
count   1936758       1247488       1247488   1247488        1247488   
mean          0            19             4        15              0   
std           0            44            21        34              2   
min           0             0             0         0              0   
25%           0             0             0         0              0   
50%           0             2             0         2              0   
75%           0            21             0        15              0   
max           1          2436          1352      1357            392   

       LateAircraftDelay  
count            1247488  
mean                  25  
std                   42  
min                    0  
25%                    0  
50%                    8  
75%                   33  
max                 1316

### **_Data Exploration_**

The data has 1,936,758 rows and 30 columns. This is a data for Airlines delay statistics from the year 2008, and includes columns such as *departure time*, *arrival time*, *flight distance*, if the flights were *diverted and/or cancelled* and so on. Some of the columns may not be useful for the analysis concerning flight delays. Let's first see what are all the columns and the datatype for each of them. The built-in function of pandas called "*data.info*" will give us this information. 

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 30 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Unnamed: 0         int64  
 1   Year               int64  
 2   Month              int64  
 3   DayofMonth         int64  
 4   DayOfWeek          int64  
 5   DepTime            float64
 6   CRSDepTime         int64  
 7   ArrTime            float64
 8   CRSArrTime         int64  
 9   UniqueCarrier      object 
 10  FlightNum          int64  
 11  TailNum            object 
 12  ActualElapsedTime  float64
 13  CRSElapsedTime     float64
 14  AirTime            float64
 15  ArrDelay           float64
 16  DepDelay           float64
 17  Origin             object 
 18  Dest               object 
 19  Distance           int64  
 20  TaxiIn             float64
 21  TaxiOut            float64
 22  Cancelled          int64  
 23  CancellationCode   object 
 24  Diverted           int64  
 25  CarrierDelay      

Now, let's add a column called *'DepDate'* with Year-Month-Day format. This way I can remove the columns of 'Year' and 'DayofMonth'.

In [4]:
# 'Departure_date' with Year-Month-Day format
data['DepDate'] = pd.to_datetime(data.Year*10000+data.Month*100+data.DayofMonth,format='%Y%m%d')

### **_Removal of columns_**

Now, Considering that the idea is to have an analysis for flight delays, the following columns can be deleted.  

    Unnamed: 0 - empty, just an index column
    Year - converted to date
    DayofMonth - converted to date
    DepTime - DepDelay is more useful
    ArrTime, CRSArrTime - ArrDelay is more useful
    FlightNum - carrier informations is more
    TailNum - a unique value, not useful, besides carrier information is more useful  
    ActualElapsedTime - it is a flight duration in minutes, does not impact flight delays. I will use flight distance instead of this to see the impact on flight delays, which is more robust.
    CRSElapsedTime - scheduled flight duration, again distance is a more robust parameter
    TaxiIn, TaxiOut - Taxi time before departure and after arrival, do not appear to impact significantly to flight delays in general. Also, 'AirTime' seems a slightly more useful factor to understand flight delays. 

In [5]:
drop_columns = ['Unnamed: 0', 'Year', 'DayofMonth', 'DepTime', 'ArrTime', 'CRSArrTime', 
                'FlightNum', 'TailNum', 'ActualElapsedTime', 'CRSElapsedTime', 'TaxiIn', 'TaxiOut']
#data_new = data
data.drop(drop_columns, axis = 1, inplace = True)

In [6]:
print('Shape of the reduced data is', data.shape)
data.head(4)

Shape of the reduced data is (1936758, 19)


Month  DayOfWeek  CRSDepTime UniqueCarrier  AirTime  ArrDelay  DepDelay  \
0      1          4        1955            WN      116       -14         8   
1      1          4         735            WN      113         2        19   
2      1          4         620            WN       76        14         8   
3      1          4        1755            WN       77        34        34   

  Origin Dest  Distance  Cancelled CancellationCode  Diverted  CarrierDelay  \
0    IAD  TPA       810          0                N         0           nan   
1    IAD  TPA       810          0                N         0           nan   
2    IND  BWI       515          0                N         0           nan   
3    IND  BWI       515          0                N         0             2   

   WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay    DepDate  
0           nan       nan            nan                nan 2008-01-03  
1           nan       nan            nan                nan 2008-01-03  
2           nan       nan            nan                nan 2008-01-03  
3             0         0              0                 32 2008-01-03

    After removing the columns that appear to be of no use, and adding a column for departure date, we have 18 columns in the dataset.

## Exercise 2

Make a complete report of the date set : 
     
     Summarize the columns of interest statistically
     Find how many missing data are per column
     Create new columns (average flight speed, whether late or not ...)
     Table of airlines with the most accumulated delays
     What are the longest flights? And the most delay? Etc.

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1936758 entries, 0 to 1936757
Data columns (total 19 columns):
 #   Column             Dtype         
---  ------             -----         
 0   Month              int64         
 1   DayOfWeek          int64         
 2   CRSDepTime         int64         
 3   UniqueCarrier      object        
 4   AirTime            float64       
 5   ArrDelay           float64       
 6   DepDelay           float64       
 7   Origin             object        
 8   Dest               object        
 9   Distance           int64         
 10  Cancelled          int64         
 11  CancellationCode   object        
 12  Diverted           int64         
 13  CarrierDelay       float64       
 14  WeatherDelay       float64       
 15  NASDelay           float64       
 16  SecurityDelay      float64       
 17  LateAircraftDelay  float64       
 18  DepDate            datetime64[ns]
dtypes: datetime64[ns](1), float64(8), int64(6), object(4)
memory us

### **_Statistical summary of columns_**

I chose following columns where statistical summary may give useful information. 
    
    - AirTime, ArrDelay, DepDelay and Distance
    - CarrierDelay, WeatherDelay, NASDelay, SecurityDelay, LateAircraftDelay

In [8]:
col_of_interest = ['AirTime', 'ArrDelay', 'DepDelay', 'Distance', 'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay', 'LateAircraftDelay']
data[col_of_interest].describe()

AirTime  ArrDelay  DepDelay  Distance  CarrierDelay  WeatherDelay  \
count  1928371   1928371   1936758   1936758       1247488       1247488   
mean       108        42        43       766            19             4   
std         69        57        53       574            44            21   
min          0      -109         6        11             0             0   
25%         58         9        12       338             0             0   
50%         90        24        24       606             2             0   
75%        137        56        53       998            21             0   
max       1091      2461      2467      4962          2436          1352   

       NASDelay  SecurityDelay  LateAircraftDelay  
count   1247488        1247488            1247488  
mean         15              0                 25  
std          34              2                 42  
min           0              0                  0  
25%           0              0                  0  
50%           2              0                  8  
75%          15              0                 33  
max        1357            392               1316

The following observations can be made from this summary:

    - The average time for flights in air is 108 min and the maximum flight time is 1091 minutes. 
    - The average distance travelled is 766 miles, and the maximum distance is 4962 miles. 
    - The average arrival and departure delays are 42 and 43 minutes, respectively. 
    - The minimum times for departure delay is 6 minutes and for arrival delay is -109 minutes, the later is puzzling because it indicates that the flight arrived 109 minutes before scheduled arrival time. There are also other negative arrival delays in the column. 
    - From the remaining columns, we can see that the carrier delays (due to airline) is a major reason for delays (2436), while security related delays lead to least number of flight delays (392). 

### **_Missing data per columns_**

To find out how many missing values are present in the data, the built-in function of pandas called "*isnull*" is used. Adding the number of times a missing data is detected over all the rows and columns, we can get total number of missing values. To get the missing values in each column, the detected values are summed once (only over columns).

In [9]:
nan_info=pd.DataFrame(data.dtypes).T.rename(index={0:'column type'})

nan_info=nan_info.append(pd.DataFrame(data.isnull().sum()).T.rename(index={0:'null values (nb)'}))

nan_info=nan_info.append(pd.DataFrame(data.isnull().sum()/data.shape[0]*100)
                         .T.rename(index={0:'null values (%)'}))
nan_info.T

column type null values (nb) null values (%)
Month                       int64                0               0
DayOfWeek                   int64                0               0
CRSDepTime                  int64                0               0
UniqueCarrier              object                0               0
AirTime                   float64             8387               0
ArrDelay                  float64             8387               0
DepDelay                  float64                0               0
Origin                     object                0               0
Dest                       object                0               0
Distance                    int64                0               0
Cancelled                   int64                0               0
CancellationCode           object                0               0
Diverted                    int64                0               0
CarrierDelay              float64           689270              36
WeatherDelay              float64           689270              36
NASDelay                  float64           689270              36
SecurityDelay             float64           689270              36
LateAircraftDelay         float64           689270              36
DepDate            datetime64[ns]                0               0

    - The columns of "AirTime" and "ArrDelay" has 8387 missing values. 
    - The columns of "CarrierDelay", "WeatherDelay", "NASDelay", "SecurityDelay" and "LateAircraftDelay" have      689270 missing values. 
    - Ideally, the missing values should be treated in a way that is appropriate for the kind of data we have. There are several ways to do it, but we will not do it at the moment. We may look at it in more detail in later parts of the course. 

### **_Creating New columns_**

Now, let's add some new columns to the data to give some extra information regarding flight delay statistics.

1. *AvgSpeed*: Average flight speed, calculated with distance divided by airtime (converted to miles/hour)
2. *Delay*: Delay status: on time (0), delayed (1). I have considered a flight to be **late only if the delay is more than 5 mins**. This is to keep a buffer time of 5 minutes, as well as to consider the possible error in the data for small delays.
3. *FLstatus*: Flight status, 
        - on time (0) if the flight has less than 5 min of arrival delay 
        - slightly delayed (1) if the flight has arrival delay is between 5 and 60 minutes
        - highly delayed (2) if the flight has arrival delay of more than 60 minutes
        - diverted (3) if the flight has been diverted
        - cancelled (4) if the flight has been cancelled
    

In [10]:
data['AvgSpeed'] = data['Distance'] / data['AirTime']*60 # the unit is miles/hour
#data.head(1)

In [11]:
for dataset in data:
    data.loc[data['ArrDelay'] <= 5, 'Delay'] = 0
    data.loc[data['ArrDelay'] >= 5, 'Delay'] = 1

In [12]:
for dataset in data:
    data.loc[data['ArrDelay'] <= 5, 'FLstatus'] = 0
    data.loc[data['ArrDelay'] >= 5, 'FLstatus'] = 1
    data.loc[data['ArrDelay'] >= 60, 'FLstatus'] = 2
    data.loc[data['Diverted'] == 1, 'FLstatus'] = 3
    data.loc[data['Cancelled'] == 1, 'FLstatus'] = 4

In [13]:
print('Shape of the data with added columns is', data.shape)
data.head(4)

Shape of the data with added columns is (1936758, 22)


Month  DayOfWeek  CRSDepTime UniqueCarrier  AirTime  ArrDelay  DepDelay  \
0      1          4        1955            WN      116       -14         8   
1      1          4         735            WN      113         2        19   
2      1          4         620            WN       76        14         8   
3      1          4        1755            WN       77        34        34   

  Origin Dest  Distance  Cancelled CancellationCode  Diverted  CarrierDelay  \
0    IAD  TPA       810          0                N         0           nan   
1    IAD  TPA       810          0                N         0           nan   
2    IND  BWI       515          0                N         0           nan   
3    IND  BWI       515          0                N         0             2   

   WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay    DepDate  \
0           nan       nan            nan                nan 2008-01-03   
1           nan       nan            nan                nan 2008-01-03   
2           nan       nan            nan                nan 2008-01-03   
3             0         0              0                 32 2008-01-03   

   AvgSpeed  Delay  FLstatus  
0       419      0         0  
1       430      0         0  
2       407      1         1  
3       401      1         1

    Now that we have converted "Diverted" and "Cancelled" to "FLstatus", we can delete these two columns. 

In [14]:
data = data.drop("Cancelled",1)
data = data.drop("Diverted",1)

In [15]:
print('Shape of the data is', data.shape)
data.head(4)

Shape of the data is (1936758, 20)


Month  DayOfWeek  CRSDepTime UniqueCarrier  AirTime  ArrDelay  DepDelay  \
0      1          4        1955            WN      116       -14         8   
1      1          4         735            WN      113         2        19   
2      1          4         620            WN       76        14         8   
3      1          4        1755            WN       77        34        34   

  Origin Dest  Distance CancellationCode  CarrierDelay  WeatherDelay  \
0    IAD  TPA       810                N           nan           nan   
1    IAD  TPA       810                N           nan           nan   
2    IND  BWI       515                N           nan           nan   
3    IND  BWI       515                N             2             0   

   NASDelay  SecurityDelay  LateAircraftDelay    DepDate  AvgSpeed  Delay  \
0       nan            nan                nan 2008-01-03       419      0   
1       nan            nan                nan 2008-01-03       430      0   
2       nan            nan                nan 2008-01-03       407      1   
3         0              0                 32 2008-01-03       401      1   

   FLstatus  
0         0  
1         0  
2         1  
3         1

    Now, renaming the columns of 'UniqueCarrier' to 'Airline' and 'Dest' to 'Destination'. 

In [16]:
data.rename(columns={'UniqueCarrier': 'Airline'}, inplace=True)
data.rename(columns={'Dest': 'Destination'}, inplace=True)
data.head(2)

Month  DayOfWeek  CRSDepTime Airline  AirTime  ArrDelay  DepDelay Origin  \
0      1          4        1955      WN      116       -14         8    IAD   
1      1          4         735      WN      113         2        19    IAD   

  Destination  Distance CancellationCode  CarrierDelay  WeatherDelay  \
0         TPA       810                N           nan           nan   
1         TPA       810                N           nan           nan   

   NASDelay  SecurityDelay  LateAircraftDelay    DepDate  AvgSpeed  Delay  \
0       nan            nan                nan 2008-01-03       419      0   
1       nan            nan                nan 2008-01-03       430      0   

   FLstatus  
0         0  
1         0

###  **_Most accumulated delays_**

Table of airlines with the most accumulated delays. 
    
    0 - total number of not delayed flights 
    1 - total number of delayed flights 


In [17]:
group_by_airline = data.groupby(['Airline','Delay'])

In [18]:
count_delays_by_airline = group_by_airline.size().unstack()
count_delays_by_airline

Delay        0       1
Airline               
9E        7622   43947
AA       28881  162029
AQ         153     591
AS        7395   31615
B6        9449   45476
CO       23313   76418
DL       19888   93840
EV       11061   70701
F9        4846   23378
FL       10294   60675
HA         722    6750
MQ       18726  122497
NW       10485   68358
OH        5469   46984
OO       18033  113747
UA       24710  116194
US       21252   76755
WN       86089  290112
XE       14530   88617
YV        6113   60656

### **_Longest flights_**

What are the longest flights?

    To check which are the longest flights in our data, let's use the "loc", which is a location identifier for    pandas dataframes. I have used "Distance" column as the criteria to find the longest flights. 

In [19]:
long_flights = data.loc[data['Distance'] == data['Distance'].max()]
long_flights

Month  DayOfWeek  CRSDepTime Airline  AirTime  ArrDelay  DepDelay  \
173177       1          1        1320      CO      605       -15        11   
173676       1          2        1320      CO      617        12        14   
173695       1          7        1320      CO      621        -8         6   
173822       1          5        1320      CO      629        27        24   
174075       1          7        1320      CO      616        -5        10   
...        ...        ...         ...     ...      ...       ...       ...   
1930954     12          6        1335      CO      609       -24         9   
1931031     12          5        1335      CO      591         2        53   
1931367     12          3        2055      CO      536        14        10   
1931368     12          3        1335      CO      599        30        66   
1931556     12          2        1335      CO      610        38        60   

        Origin Destination  Distance CancellationCode  CarrierDelay  \
173177     EWR         HNL      4962                N           nan   
173676     EWR         HNL      4962                N           nan   
173695     EWR         HNL      4962                N           nan   
173822     EWR         HNL      4962                N            24   
174075     EWR         HNL      4962                N           nan   
...        ...         ...       ...              ...           ...   
1930954    EWR         HNL      4962                N           nan   
1931031    EWR         HNL      4962                N           nan   
1931367    HNL         EWR      4962                N           nan   
1931368    EWR         HNL      4962                N             0   
1931556    EWR         HNL      4962                N            38   

         WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay    DepDate  \
173177            nan       nan            nan                nan 2008-01-14   
173676            nan       nan            nan                nan 2008-01-15   
173695            nan       nan            nan                nan 2008-01-13   
173822              0         3              0                  0 2008-01-11   
174075            nan       nan            nan                nan 2008-01-06   
...               ...       ...            ...                ...        ...   
1930954           nan       nan            nan                nan 2008-12-06   
1931031           nan       nan            nan                nan 2008-12-05   
1931367           nan       nan            nan                nan 2008-12-03   
1931368             0         0              0                 30 2008-12-03   
1931556             0         0              0                  0 2008-12-02   

         AvgSpeed  Delay  FLstatus  
173177        492      0         0  
173676        483      1         1  
173695        479      0         0  
173822        473      1         1  
174075        483      0         0  
...           ...    ...       ...  
1930954       489      0         0  
1931031       504      0         0  
1931367       555      1         1  
1931368       497      1         1  
1931556       488      1         1  

[324 rows x 20 columns]

In [20]:
ewr = len(long_flights.loc[long_flights['Origin'] != "EWR"])
hnl = len(long_flights.loc[long_flights['Origin'] != "HNL"])
print(ewr + hnl)

324


    I found 324 flights, all of which are of long distance (4962 miles) and are operating between "EWR" and "HNL".
    
Now, let's check which flights have a long airtime. 

In [21]:
data.loc[data['AirTime'] == data['AirTime'].max()]

Month  DayOfWeek  CRSDepTime Airline  AirTime  ArrDelay  DepDelay  \
1488690      9          2         920      HA     1091      1050       291   

        Origin Destination  Distance CancellationCode  CarrierDelay  \
1488690    SEA         HNL      2677                N          1040   

         WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay    DepDate  \
1488690             0         0              0                 10 2008-09-09   

         AvgSpeed  Delay  FLstatus  
1488690       147      1         2

    In case of long flights by 'Airtime', I get the flight from "SEA" to "HNL" on September 9, 2008 that spent 18 hours and 11 minutes (1091 minutes) in air. However, the distance of this flight is not the largest one, but 2677 miles. 
    
    Looking at arrival delay (ArrDelay), the flight seems to be late by 1050 minutes (i.e., 17 hours, 30 minutes), of which delay due to previous flight being delayed (LateAircraftDelay) was 10 minutes. Also note that the carrier delay for this flight is 17 hours and 19 minutes (1040 minutes), indicating that the long arrival delay was almost entirely caused by the carrier (i.e., airline).

### **_Longest delay_**

Which flight has the most delay?

    Let's get this information by getting the information of the flight with the longest arrival delay (ArrDelay).

In [25]:
data.loc[data['ArrDelay'] == data['ArrDelay'].max()]

Month  DayOfWeek  CRSDepTime Airline  AirTime  ArrDelay  DepDelay  \
322516      2          7        1820      NW      437      2461      2457   

       Origin Destination  Distance CancellationCode  CarrierDelay  \
322516    HNL         MSP      3972                N          1455   

        WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay    DepDate  \
322516             0         4              0               1002 2008-02-03   

        AvgSpeed  Delay  FLstatus  
322516       545      1         2

    The flight with the most delay, that of 41 hours (2461 minutes), is the one between "HNL" and "MSP", on February 3, 2008 operated by Airline "NW". The distance between these two airports is 3972 miles. Of this 41 hours delay, the previous flight's late arrival caused 16 hours and 42 minutes, carrier related delay was 24 hours and 15 minutes and National Air System Delay (NASDelay) caused 4 minutes.

## Exercise 3

Export the data set clean and with the new columns to Excel.

    The dataset is too large to export to excel, therefore, I will export it to csv. 

In [26]:
data.to_csv('../Data/FlightDelay_new.csv')  